### ISSUE TITLE
Simplify Code Base

### ISSUE THREAD

**Issue by climba03003** [fastify/fastify#4452]
### Prerequisites

- [X] I have written a descriptive issue title
- [X] I have searched existing issues to ensure the feature has not already been requested

### 🚀 Feature Proposal

The current code base is complicated and may not be good to maintain in long terms.
This issue would be a long terms goal and discussion.

I think we can do something in short. For example.
- [ ] Extract `lib/server.js` in separate module
The reason behind is simple, it is a complicated startup for multiple server. By extracting into a single module, we can control the initialization of `http` server better (e.g. Deprecation of variadic signature)

In long terms
- [ ] Better comments for workflow and JSDoc
Currently, it is hard to follow how `request` and `response` works in code. The problem is that it will increase the debug time for issue. And have a burden for new contributors.

**Comment by gurgunday** [comment]
I know this is probably not the form of simplification you’re talking about, but there was some talk on removing the non-async, callback-based API for handlers.

I think that would really help new contributors. Maybe for the long-term?

**Comment by climba03003** [comment]
> I know this is probably not the form of simplification you’re talking about, but there was some talk on removing the non-async, callback-based API for handlers.

When you are referencing `some talk`, it is better to provide the link of the talk.
It helps us know what exactly you are talking about and check the pros and cons through the talk.

**Comment by mcollina** [comment]
I'm +1 with @climba03003 plan.

I don't think we should drop callbacks. They are extremely useful to handle some edge cases. It's such a massive endeavor that's close to a big bang rewrite.

**Comment by mcollina** [comment]
Something we should refactor is support some kind of check to throw an error when a plugin needs to do some metadata check (like onRoute) and it's not awaited.

**Comment by gurgunday** [comment]
> > I know this is probably not the form of simplification you’re talking about, but there was some talk on removing the non-async, callback-based API for handlers.
> 
> When you are referencing `some talk`, it is better to provide the link of the talk. It helps us understand what exactly you are talking about and check the pros and cons through the talk.

Yeah you’re right — sorry I was (and still am) in class right now. But that didn’t prevent me from finding [one occurrence](https://github.com/fastify/fastify-compress/pull/224#issuecomment-1118106465)  of that talk 🤣

**Comment by jsumners** [comment]
> * Extract `lib/server.js` in separate module

Do you mean a literal `npmjs.com` module, e.g. `@fastify/server`? If so, I disagree. That would make maintenance more difficult, and increase the complexity of the code base.

> * Better comments for workflow and JSDoc

This I agree with. You may have noticed, but I have been requiring JSDoc in all of the code that I have been reviewing that adds new functions.

**Comment by cesarvspr** [comment]
I will be happy to work on it @climba03003 

Would you mind providing a clarification about what you mean by ` Extract lib/server.js in separate module`?

**Comment by Uzlopak** [comment]
@cesarvspr 

We have to first discuss, if this is the way to go or not. Also this is not a first-good-issue. I dont think that this is a simple patch.

**Comment by cesarvspr** [comment]
> @cesarvspr 
> 
> We have to first discuss, if this is the way to go or not. Also this is not a first-good-issue. I dont think that this is a simple patch. 

Yes @Uzlopak you are right, also I appreciate the warning. Just in case this issue has room to me I will be happy to contribute. Otherwise I will understand.

**Comment by metcoder95** [comment]
Just adding my tiny two cents, I agree 100% with the points about adding JSDoc across the codebase, most important on those points where a variable is set in a previous point and there are several dependencies down-level and across callbacks within the same function or even script.

Along with the previous point, one of the things that I've seen more friction (and I also have struggled to draw a mental map of it), is how the loading of plugins/routes works in mix with `avvio`.
That can also be accompanied by @mcollina point of handling better those plugins that require to be `await`ed to work (e.g. those ones that set an `onRoute` hook), which has been subjected to quite a good number of opened issues due to plugins not working due to not waiting for them to be loaded.

**Comment by erfanium** [comment]
I'm also in favor of dropping for callback support for next fastify major versions. I found fastify internal code hard to read, because of this unnecessary complexity.

there's always a `done` parameter passing from a function to another, very hard to find out what is really going on there.
